### ISSUE TITLE
[Usage]: How to evaluate a Pi0 based model?

### ISSUE THREAD

**Issue by TwSphinx54** [RLinf/RLinf#290]
### Problem description

When I use `bash examples/embodiment/eval_embodiment.sh libero_spatial_ppo_openpi_pi05_eval` to evaluate the Pi0-based model, the server's CPU memory quickly climbs to its peak, then the IO rapidly maxes out, ultimately causing the server to become unresponsive. I want to know if there's an issue with my usage of the evaluation script or if the related configuration is incorrect. Or is the evaluation of Pi0 in the paper conducted using the method from [openpi](https://github.com/Physical-Intelligence/openpi)? 

[libero_spatial_ppo_openpi_pi05_eval.yaml](https://github.com/user-attachments/files/23585432/libero_spatial_ppo_openpi_pi05_eval.yaml)

Besides, I have set:
```
export MUJOCO_GL="egl"
export PYOPENGL_PLATFORM="egl"
```

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by LiuZhihao2022** [comment]
Hi! The issue is caused by the setting `num_envs: 500` in  
`examples/embodiment/config/env/eval/libero_spatial.yaml`.  
This value determines how many environments are launched **in parallel** during evaluation.  

We recommend the following steps:

## 1. Reduce `num_envs`

Modify in `examples/embodiment/config/env/eval/libero_spatial.yaml`:

```yaml
num_envs: <a smaller value>
```

## 2. Increase `n_eval_chunk_steps`  

Modify in your `libero_spatial_ppo_openpi_pi05_eval.yaml`:

```yaml
n_eval_chunk_steps: <corresponding larger value>
```

### Keep the ratio consistent  

If you reduce `num_envs` by **1/2**, you should increase `n_eval_chunk_steps` by **2×**.  
These two parameters scale inversely to preserve correct evaluation behavior.

---

Additionally, we noticed that you modified the default value of `n_eval_chunk_steps`.  
By default:

```
n_eval_chunk_steps = max_episode_steps / num_action_chunks
```

Here is a recommended configuration:

**In `examples/embodiment/config/env/eval/libero_spatial.yaml`**

```yaml
num_envs: 125    # one quarter of the original default
```

**In your `libero_spatial_ppo_openpi_pi05_eval.yaml`**

```yaml
n_eval_chunk_steps: 192    # 4× the original default
```

---

## Additional Recommendation for Evaluation

Because launching too many environments simultaneously can put heavy load on the GPU and rendering backend,  
we also strongly recommend enabling offscreen CPU-based rendering during evaluation:

```bash
export MUJOCO_GL="osmesa"
export PYOPENGL_PLATFORM="osmesa"
```

This greatly reduces GPU memory usage and avoids GL context conflicts when running many environments in parallel.

---

Let us know if this resolves your issue!

**Comment by TwSphinx54** [comment]
Thanks a lot for your detailed explanation! It perfectly works for me.
As my GPUs have enough mem, `egl` is still applied.

Following your suggestion, my config is (slightly adjusted for a safer reservation):
```
num_envs: 100
n_eval_chunk_steps: 240
```
For reference, the evaluation consumes:
- GPU mem: 24.2G
- RAM: 218G

Thank you again for your great work!

**Comment by GostInShell** [comment]
@LiuZhihao2022 I run the results with below parameters following your recommendation.
```
num_envs: 125    # one quarter of the original default
n_eval_chunk_steps: 192    # 4× the original default
```
But I notice that the generated video length becomes four times of the original. However, the `state_id` seems remain the **same**, i.e., the initial setup is the same. Are there other parameters should be changed to avoid this?

Moreover, I doubt whether `success_once` still represent the overall success rate now that it runs four episodes in one video now. (Does that mean `success_once` become one if any of the four episodes in one video success?)

**Comment by LiuZhihao2022** [comment]
@GostInShell Thanks for your question, and for trying those settings!

You’re right — this is actually a bug. During evaluation, each reset is supposed to switch to a different `reset_state_id`, but at the moment all four episodes start from the same state, which is why the video becomes four times longer but looks identical in the beginning.

About `success_once`:
It still reflects the overall success rate. Each of the four episodes is counted independently, so if any of them succeeds, it contributes to the metric. It’s not just checking success once per video.

We’ll fix the reset_state_id switching behavior in an upcoming update — thanks a lot for pointing it out!

**Comment by YangYang-DLUT** [comment]
Follow this lead, I evaluated the ``RLinf-OpenVLAOFT-GRPO-LIBERO-spatial`` model, and get 
``'eval/success_once': array(0.81875, dtype=float32)``, lower than reported here https://huggingface.co/RLinf/RLinf-OpenVLAOFT-GRPO-LIBERO-spatial
my `libero_spatial_grpo_openvlaoft_eval.yaml`：
```
defaults:
  - env/train: libero_spatial
  - env/eval: libero_spatial
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor,env,rollout: all

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openvla"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  only_eval: True
  val_check_interval: -1
  save_interval: 25
  seq_length: 4096
  max_prompt_length: 128

algorithm:
  auto_reset: False
  ignore_terminations: False
  use_fixed_reset_state_ids: True
  specific_reset_id: null  # ← 添加这一行
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 8
  n_chunk_steps: ${int_div:${env.train.max_episode_steps}, ${actor.model.num_action_chunks}}
  # n_eval_chunk_steps: ${int_div:${env.train.max_episode_steps}, ${actor.model.num_action_chunks}}
  n_eval_chunk_steps: 1280
  num_group_envs: 8
  rollout_epoch: 8
  reward_type: action_level  # action_level or chunk_level
  logprob_type: token_level
  entropy_type: token_level

  adv_type: grpo
  loss_type: actor
  loss_agg_func: "token-mean"
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.28
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  reward_coef: 5.0
  max_episode_steps: 512
  use_ordered_reset_state_ids: True

  sampling_params:
    use_greedy: False
    temperature_train: 1.6
    temperature_eval: 1.6
    top_k: -1
    top_p: 1.0
    repetition_penalty: 1.0

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 1024
    min_length: 1

  filter_rewards: True
  rewards_lower_bound: 0.5
  rewards_upper_bound: 4.5

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  generation_backend: "huggingface"
  model_dir: "/user/hadoop-mlm-hl/hadoop-mlm/yangyang260/proj/RLinf/ckpts/RLinf-OpenVLAOFT-GRPO-LIBERO-spatial"
  precision: ${actor.model.precision}
  enable_offload: False
  pipeline_stage_num: 1

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  checkpoint_load_path: "/user/hadoop-mlm-hl/hadoop-mlm/yangyang260/proj/RLinf/ckpts/RLinf-OpenVLAOFT-GRPO-LIBERO-spatial"
  checkpoint_save_path: "../results"
  micro_batch_size: 32
  global_batch_size: 8192
  seed: 1234
  enable_offload: False

  model:
    model_name: "openvla_oft"
    ckpt_path: null
    value_type: ${algorithm.reward_type}  # 'action' or 'token'
    action_dim: 7
    num_action_chunks: 8
    use_proprio: False
    unnorm_key: libero_spatial_no_noops
    center_crop: True

    precision: "bf16"
    add_bias_linear: False
    add_qkv_bias: True
    vocab_size: 32000
    hidden_size: 4096
    add_value_head: False
    image_size: [224, 224]
    is_lora: False
    lora_rank: 32
    use_wrist_image: False
    attn_implementation: "flash_attention_2"
    low_cpu_mem_usage: True
    trust_remote_code: True
    gradient_checkpointing: False

  optim:
    lr: 2.0e-5
    value_lr: 3.0e-3
    adam_beta1: 0.9
    adam_beta2: 0.999
    adam_eps: 1.0e-05
    clip_grad: 1.0

  tokenizer:
    tokenizer_type: "HuggingFaceTokenizer"
    tokenizer_model: "/user/hadoop-mlm-hl/hadoop-mlm/yangyang260/proj/RLinf/ckpts/RLinf-OpenVLAOFT-GRPO-LIBERO-spatial"
    use_fast: False
    trust_remote_code: True
    padding_side: "right"
    
  fsdp_config:                  # below are fsdp configs, for more details, see https://pytorch.org/docs/stable/fsdp.html and https://huggingface.co/docs/accelerate/package_reference/fsdp

    strategy: "fsdp"            # FSDP strategy: ["fsdp", "fsdp2"]

    # Sharding strategy: "full_shard" (shard parameters, gradients, optimizer states),
    # "shard_grad_op" (shard gradients and optimizer states only), 
    # "hybrid_shard" (combines data parallelism and model parallelism - sharding within FSDP groups, replicate across DDP groups),
    # "no_shard" (no sharding)
    sharding_strategy: "full_shard" 

    cpu_offload: False           # whether to offload parameters and gradients to CPU when not in use，if True, actor's enable_offload should be True too.
    offload_pin_memory: False    # whether FSDP2's CPU offload policy should pin memory when cpu_offload is True
    reshard_after_forward: True  # if True, FSDP2 will reshard parameters after forward pass to save memory

    enable_gradient_accumulation: True  # if True, gradient accumulation will be enabled for FSDP/FSDP2 training, which will enhance training performance but increase memory usage.
    forward_prefetch: False             # if True, FSDP will explicitly prefetches the next upcoming all-gather while executing in the forward pass. only use with static graphs. Overlaps communication with computation to improve performance.
    limit_all_gathers: False            # if True, FSDP will synchronizes CPU threads to limit the number of concurrent all-gathers. Only affects strategies that schedule all-gathers
    backward_prefetch: null             # options are null, 'pre', 'post'. if 'pre', FSDP will prefetch the next upcoming all-gather while computing gradients. if 'post', FSDP will prefetch the next all-gather until current gradient is computed.
    use_orig_params: False              # if True, FSDP will use module's original parameters, it means it will expose nn.Module.named_parameters rather than FlatParameter
    use_liger_kernel: False             # if True, liger_kernel will be used for FSDP, note that currently supported models in RLinf are [qwen2.5, qwen2.5-vl], for more details, see liger_kernel's doc

    fsdp_size: -1               # Number of GPUs per FSDP group for hybrid sharding. -1 means use all available GPUs in a single FSDP group

    mixed_precision:            # mixed precision settings for fsdp/fsdp2
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

    amp:
      enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
      precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
      use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.

reward:
  use_reward_model: False

critic:
  use_critic_model: False
```

my `libero_spatial.yaml`:
```
simulator_type: libero
task_suite_name: libero_spatial

auto_reset: True
ignore_terminations: True
max_episode_steps: ${algorithm.max_episode_steps}

use_rel_reward: True
reward_coef: ${algorithm.reward_coef}
only_eval: True

seed: 0
num_group: ${env.eval.num_envs}
group_size: 1
use_fixed_reset_state_ids: True
use_wrist_image: ${actor.model.use_wrist_image}

num_envs: 25
specific_reset_id: ${env.train.specific_reset_id}

video_cfg:
  save_video: True
  info_on_video: True
  video_base_dir: ${runner.logger.log_path}/video/eval

init_params:
  camera_heights: 256
  camera_widths: 256
```

Any clues why this diff on the perfomance?

**Comment by LiuZhihao2022** [comment]
> [@GostInShell](https://github.com/GostInShell) Thanks for your question, and for trying those settings!
> 
> You’re right — this is actually a bug. During evaluation, each reset is supposed to switch to a different `reset_state_id`, but at the moment all four episodes start from the same state, which is why the video becomes four times longer but looks identical in the beginning.
> 
> About `success_once`: It still reflects the overall success rate. Each of the four episodes is counted independently, so if any of them succeeds, it contributes to the metric. It’s not just checking success once per video.
> 
> We’ll fix the reset_state_id switching behavior in an upcoming update — thanks a lot for pointing it out!

@YangYang-DLUT I'm sorry for this. As mentioned here, there is a bug in evaluation and we have fixed it in [pr363](https://github.com/RLinf/RLinf/pull/363). You can pull the latest version of code and try it again.
